File: Leetcode/23/23.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

struct Node {
    int val;
    int pos;
    Node() = default;
    Node(int val, int pos) : val(val), pos(pos) {}
    bool operator > (const Node& node) const {
        return val > node.val;
    }
};

class Solution {
public:
    // k-way merge with minHeap in O(k) + O(nlogk)
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        priority_queue<Node, vector<Node>, greater<>> pq;
        int n = (int)lists.size();
        for (int i = 0; i < n; ++i) {
            if (lists[i]) {
                pq.push(Node(lists[i]->val,i));
            }
        }
        ListNode* head = new ListNode();
        ListNode* currentNode = nullptr;
        Node node;
        while (!pq.empty()) {
            node = pq.top();
            pq.pop();
            if (currentNode) {
                auto temp = new ListNode(node.val);
                currentNode->next = temp;
                currentNode = temp;
            }
            else {
                currentNode = new ListNode(node.val);
                head->next = currentNode;
            }
            if (lists[node.pos]->next) {
                lists[node.pos] = lists[node.pos]->next;
                pq.push(Node(lists[node.pos]->val, node.pos));
            } 
        }
        return head->next;
    }
};

File: Leetcode/23/23.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
const MAX_VALUE = 0x3f3f3f3f

// O(k*max(len(lists[i]))
func mergeKLists(lists []*ListNode) *ListNode {
	// Corner case
	if len(lists) == 0 {
		return nil
	}
    head := new(ListNode)
	var currentNode *ListNode = nil
	// Collect all nodes such that each node is the header of each linked list
	var nodes = make([]*ListNode, len(lists))
	for i := 0; i < len(lists); i++ {
		nodes[i] = lists[i]
	}
	for {
		// Scan again if any node in the list of linked list does not reach the end
		scanAgain := false
		temp := make([]int, len(nodes))
 		for i := 0; i < len(nodes); i++ {
			if nodes[i] == nil {
				temp[i] = MAX_VALUE
			} else {
				scanAgain = true
				temp[i] = nodes[i].Val
			}
		}
		if !scanAgain {
			break
		}
		minVal := MAX_VALUE
		minValIndex := 0
		for i := 0; i < len(temp); i++ {
			if temp[i] < minVal {
				minVal = temp[i]
				minValIndex = i
			}
		}
		if currentNode == nil {
			currentNode = new(ListNode)
			currentNode.Val = minVal
	        head.Next = currentNode
			currentNode.Next = nil
		} else {
			tempNode := new(ListNode)
			tempNode.Val = minVal
			tempNode.Next = nil
			currentNode.Next = tempNode
			currentNode = tempNode
		}
		nodes[minValIndex] = nodes[minValIndex].Next
	}
	return head.Next
}